hudi-client/hudi-client-common/src/main/java/org/apache/hudi/io/HoodieMergeHandle.java

```diff
@@ -147,6 +149,13 @@ public HoodieMergeHandle(HoodieWriteConfig config, String instantTime, HoodieTab
     this.preserveMetadata = true;
     init(fileId, this.partitionPath, dataFileToBeMerged);
     validateAndSetAndKeyGenProps(keyGeneratorOpt, config.populateMetaFields());
+    // if the old schema equals to the new schema, avoid heavy rewriting
+    if (config.populateMetaFields() && useWriterSchemaForCompaction) {
+      LOG.info("Using update instead rewriting during compaction");
+      copyOldFunc = (key, record, schema, prop) -> this.updateMetadataToOldRecord(key, record, schema, prop);
```

Comment:
So you are assuming the compaction does not involve schema evolution or just the latest schema might be okay because the schema is compatible? How can we ensure the schema compatibility?

Reply:
Not exactly.
The behavior is consistent with the old behavior.
<img width="1000" alt="image" src="https://github.com/apache/hudi/assets/1525333/2ab0c5e8-d22e-41a2-bb73-e541b64a8083">
In the above picture, if `config.populateMetaFields() && useWriterSchemaForCompaction` is true, the `oldSchema` equals to `newSchema`.

Reply:
but it still uses the latest schema as the write schema, how about the schema already evolved?

Reply:
Good question.
The responsible of this method is only merging base record and incremental record, not including handle schema evolution. 
Handling schema evolution happens before call the `HoodieMergeHandle#write` method.
<img width="1305" alt="image" src="https://github.com/apache/hudi/assets/1525333/3a03e08b-fe2e-4da6-a788-07cbb6feeadd">
<img width="1027" alt="image" src="https://github.com/apache/hudi/assets/1525333/def1f2ee-ed97-47f8-92b6-76d45500bea7">